sstables/sstable_set.cc

```diff
@@ -768,6 +768,393 @@ sstable_set_impl::create_single_key_sstable_reader(
+
+    // Determines how many times a fragment should be taken from the same
+    // reader in order to enter gallop mode. Must be greater than one.
+    static constexpr int _gallop_mode_entering_threshold = 3;
+
+    bool in_gallop_mode() const {
+        return _gallop_mode_hits >= _gallop_mode_entering_threshold;
+    }
+
+    // Relevant only in forwarding mode (in non-forwarding mode _halted_readers is always empty).
+    // Remove all halted readers (i.e. readers which returned end-of-stream) that we know for sure
+    // won't return any more fragments. We use the sstables' max_position() metadata to make the decision.
+    // Precondition: every halted reader returned all fragments with positions <= `up_to`.
+    // See also the comment for `_halted_readers`.
+    void collect_exhausted_readers(position_in_partition_view up_to) {
+        while (!_halted_readers.empty()
+                && !_less(up_to, _halted_readers.front()->sst->max_position())) {
+            boost::range::pop_heap(_halted_readers, _halted_reader_cmp);
+            _all_readers.erase(_halted_readers.back());
+            _halted_readers.pop_back();
+        }
+    }
+
+    // Retrieve the next fragment from the reader pointed to by `it`.
+    // We assume that `it` is either in `_unpeeked_readers` (for the non-galloping case)
+    // or is equal to `_galloping_reader` (for the galloping case), and that all fragments
+    // previously returned from the reader have already been returned by operator().
+    //
+    // In the non-galloping case, the peeked reader is pushed onto the _peeked_readers heap.
+    //
+    // In the galloping case, if the reader again won with other readers, the fragment is saved
+    // to be the next batch, and we return `true` to indicate that the galloping reader has won.
+    // Otherwise, the reader is pushed onto _peeked_readers and we return `false`.
+    template <bool galloping>
+    std::conditional_t<galloping, future<bool>, future<>>
+    peek_reader(db::timeout_clock::time_point timeout, reader_iterator it) {
+        return (it->reader)(timeout).then([this, timeout, it] (mutation_fragment_opt mfo) {
+            if (mfo) {
+                if (mfo->is_partition_start()) {
+                    // This path executes only on the non-galloping case (if `it` was galloping, it must have
+                    // returned at least one fragment earlier).
+
+                    // We assume there are no partition tombstones.
+                    assert(!mfo->as_partition_start().partition_tombstone());
+
+                    auto f = _forwarded_to ? it->reader.fast_forward_to(*_forwarded_to, timeout) : make_ready_future<>();
+                    return f.then([this, timeout, it] { return peek_reader<galloping>(timeout, it); });
+                }
+
+                // We assume that the schema does not have any static columns, so there cannot be any static rows.
+                assert(!mfo->is_static_row());
```

Comment:
Why not add a condition to the factory function making this reader, which falls back to the generic reader for schemas with static rows instead? This assert will make the cluster unusable for the poor soul who adds a static row.

Reply:
This assert should never fire since I've checked there are no static columns in the schema before even creating the reader. It's doesn't rely on the user not to do something (that would be a misuse of an assert).

I can replace these asserts with `on_internal_error` or even remove them completely. They're here mostly for testing.

Reply:
Yes, I saw it later. I think you can remove.

Reply:
replaced with `on_internal_error`, I wanted to be safe